### PR TITLE
Sanitize headers nominated by the Connection header

### PR DESCRIPTION
**Pull request by abaptiste**
Description: Sanitize headers nominated by the Connection header

Risk Level: Medium
Testing: Added several test cases to verify the header manipulation.   Also executed `bazel test //test/...`
Docs Changes: N/A
Release Notes: N/A
Fixes: https://github.com/envoyproxy/envoy/issues/8623

### CODE REVIEW

File: source/common/http/utility.cc [hunk 1]
```diff
@@ -387,6 +388,115 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+
+      // reject the request if the TE header is too large
+      if (is_te_header && (nominated_heaader_value_sv.size() >= 256)) {
+        return false;
+      }
+
+      // If the TE header contains "trailers" we will reset the header to this value since it's
+      // the only permitted value
+      if (is_te_header && (nominated_heaader_value_sv.find(
+                               Http::Headers::get().TEValues.Trailers) != std::string::npos)) {
```
Comment (yanavlasov) [inline source/common/http/utility.cc]: I'll defer to other reviewers to see this check is enough. It does not account for a weird edge case where a header name contains the "trailers" keyword. I thought about it more and I think your original version using StrSplit in the for range loop was correct. The StrSplit in the for range loop does not cause memory allocation and with the length check it would be safe. If the find check is deemed to be insufficient you should use check for the trailers keyword using your original method. Sorry abot flip-flopping on this one.

File: source/common/http/http1/codec_impl.h [hunk 2]
```diff
@@ -309,6 +309,7 @@ class ConnectionImpl : public virtual Connection, protected Logger::Loggable<Log
   const uint32_t max_headers_count_;
 
   bool strict_header_validation_;
+  bool connection_header_sanitization_;
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply source/common/http/http1/codec_impl.h]: Would that mean that `strict_header_validation_` is also missing a const?

File: source/common/http/utility.cc [hunk 3]
```diff
@@ -387,6 +388,115 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+    // the connection header is removed elsewhere when the H1 request is upgraded to H2
+    if (StringUtil::CaseInsensitiveCompare()(token_sv, cv.Close) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Http2Settings) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.KeepAlive) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Upgrade)) {
+      continue;
+    }
+
+    // Build the LowerCaseString for header lookup
+    const std::string header_to_remove = std::string(token_sv);
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply source/common/http/utility.cc]: Yep.  With the multiple refactors I missed this.

File: source/common/http/utility.cc [hunk 4]
```diff
@@ -387,6 +388,115 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+      return false;
+
+    } else {
+      // Examine the value of all other nominated headers
+      nominated_header = headers.get(lcs_header_to_remove);
+    }
+
+    if (nominated_header) {
+      const HeaderString& nominated_header_value = nominated_header->value();
+      auto nominated_heaader_value_sv = nominated_header_value.getStringView();
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply source/common/http/utility.cc]: I thought I fixed this. I will double check.

File: source/common/http/utility.cc [hunk 5]
```diff
@@ -387,6 +388,115 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+
+      const bool is_te_header =
+          StringUtil::CaseInsensitiveCompare()(token_sv, Http::Headers::get().TE.get());
+
+      // reject the request if the TE header is too large
+      if (is_te_header && (nominated_heaader_value_sv.size() >= 256)) {
+        return false;
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply source/common/http/utility.cc]: Certainly

File: source/common/http/utility.cc (hunk 1 above)
Comment (abaptiste) [reply source/common/http/utility.cc]: This check should only execute if the header itself is "TE".   I removed the tokenization since it was expensive (per a comment on the old PR).  In this change, if the header is "TE" and we find "trailers" and other values (there should be a check for whether a comma exists in the header value), then we would replace the existing value of `TE: whatever, and, trailers` with just `TE: trailers`.

There was a second commit where I fixed the variable spelling error and added this check.  In any case, I'm open to what others would suggest here.

File: source/common/http/utility.cc [hunk 6]
```diff
@@ -387,6 +388,116 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+  for (const auto& token : connection_header_tokens) {
+
+    const auto token_sv = StringUtil::trim(token);
+
+    // Build the LowerCaseString for header lookup
+    const LowerCaseString lcs_header_to_remove{std::string(token_sv)};
+
+    // If the Connection token value is not a nominated header, ignore it here since
+    // the connection header is removed elsewhere when the H1 request is upgraded to H2
+    if ((lcs_header_to_remove == LowerCaseString(cv.Close)) ||
```
Comment (yanavlasov) [inline source/common/http/utility.cc]: Ah, yes ConnectionValues are std::string not LowerCaseString. Conversion from std::string to LowerCaseString may cause heap allocations. For the maximum of 10 tokens this part will do 40 allocations, copies and lower case conversions. We know that ConnectionValues are lowercase. I think a better approach would be to use (lcs_header_to_remove.get() == cv.Close) pattern, in this way a per iteration work is avoided.

File: source/common/http/utility.cc [hunk 7]
```diff
@@ -387,6 +388,116 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+        return false;
+      }
+
+      // If the TE header contains "trailers" we will reset the header to this value since it's
+      // the only permitted value
+      if (is_te_header && (nominated_header_value_sv.find(Http::Headers::get().TEValues.Trailers) !=
+                           std::string::npos)) {
+
+        // Reset the TE header to just trailers only if other values exist
+        if (nominated_header_value_sv.find(',') != std::string::npos) {
```
Comment (yanavlasov) [inline source/common/http/utility.cc]: Excellent way to avoid needless clear and copy.

File: source/common/http/utility.cc [hunk 8]
```diff
@@ -387,6 +388,116 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+
+      // reject the request if the TE header is too large
+      if (is_te_header && (nominated_header_value_sv.size() >= MAX_ALLOWED_TE_VALUE_SIZE)) {
+        ENVOY_LOG_MISC(trace, "TE header contains a value that exceeds the allowable length");
+        return false;
+      }
+
+      // If the TE header contains "trailers" we will reset the header to this value since it's
+      // the only permitted value
+      if (is_te_header && (nominated_header_value_sv.find(Http::Headers::get().TEValues.Trailers) !=
```
Comment (yanavlasov) [inline source/common/http/utility.cc]: The more I think about it, the more I agree that your original approach of looking for the "trailers" keyword was the correct one. This will fail if request has the "TE: TRAILERS" header which is allowed by the standard (values are case insensitive) and TE would be removed. It will also keep the header if request has the "TE: myabusivetrailersencoding", which I'm certain someone would figure out how to abuse. I think this should go back to the range for loop over the split nominated_header_value_sv and do the trim and lower case conversion before comparing to the "trailers" keyword. Sorry about taking your time with going back and forth on this one.

Comment (abaptiste) [reply source/common/http/utility.cc]: Your example is valid.   I updated the code and added a test case that exercises this condition.

File: source/common/http/utility.cc [hunk 9]
```diff
@@ -387,6 +388,134 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+      }
+
+      if (is_te_header) {
+        keep_header = true;
+        for (const auto& header_value : absl::StrSplit(nominated_header_value_sv, ',')) {
+
+          const absl::string_view header_sv = StringUtil::trim(header_value);
+
+          // Remove everything from TE header except "trailers"
+          if (!StringUtil::CaseInsensitiveCompare()(header_sv,
```
Comment (yanavlasov) [inline source/common/http/utility.cc]: It is very close. It can be cleaned up a bit more by removing the need for the tokens_to_remove map. Note that the condition for keeping the TE header is the presence of the "trailers" token. You can do the following:
1. leave the keep_header unchanged at the beginning of the loop
2. reverse the if statement condition inside the loop, so it it is true when the header_sv == "trailers"
3. inside the if statement set the keep_header and "break" out of the loop, since we do not care to look at the rest of the tokens
After the loop, check if TE header has to be kept and if so set its value to "trailers". With this change you can remove the if statement below at line 484 entirely.
This avoids the overhead of the building the hash map and calling the "removeTokens" method.

File: source/common/http/http1/codec_impl.cc [hunk 10]
```diff
@@ -532,6 +534,15 @@ int ConnectionImpl::onHeadersCompleteBase() {
       ENVOY_CONN_LOG(trace, "codec entering upgrade mode.", connection_);
       handling_upgrade_ = true;
     }
+  } else if (connection_header_sanitization_ && current_header_map_->Connection()) {
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: can you add release notes at docs/root/intro/version_history.rst including the flag to flip to revert to prior behavior?

File: source/common/http/utility.cc [hunk 11]
```diff
@@ -387,6 +388,123 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+
+  // Remove any headers nominated by the Connection header. The TE header
+  // is sanitized and removed only if it's empty after removing unsupported values
+  // See https://github.com/envoyproxy/envoy/issues/8623
+  const auto& cv = Http::Headers::get().ConnectionValues;
+  const auto& connection_header_value = headers.Connection()->value();
+
+  StringUtil::CaseUnorderedSet headers_to_remove{};
+  std::vector<std::string> connection_header_tokens =
+      absl::StrSplit(connection_header_value.getStringView(), ',');
```
Comment (alyssawilk) [inline source/common/http/utility.cc]: can we use StringUtil::splitToken, which auto-trims, and also allows removing empty headers?

File: source/common/http/utility.cc [hunk 12]
```diff
@@ -387,6 +388,123 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+      ENVOY_LOG_MISC(trace, "Skipping self nominated header [{}]", token_sv);
+      keep_header = true;
+      headers_to_remove.emplace(token_sv);
+
+    } else if ((lcs_header_to_remove == Http::Headers::get().ForwardedFor) ||
+               (lcs_header_to_remove == Http::Headers::get().ForwardedHost) ||
+               (lcs_header_to_remove == Http::Headers::get().ForwardedProto) ||
+               !token_sv.find(':')) {
+      // If pseudo headers or X-Forwarded* headers are nominated, this could be
+      // an invalid request. Reject the request
```
Comment (alyssawilk) [inline source/common/http/utility.cc]: can we comment for posterity why we disallow these?  If it's in the spec, we can link.   If we just think it's high risk we can say that too.

File: source/common/http/utility.cc [hunk 13]
```diff
@@ -387,6 +388,123 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+      return false;
```
Comment (alyssawilk) [inline source/common/http/utility.cc]: should we trace log this as well for consistency?

File: source/common/http/utility.cc [hunk 14]
```diff
@@ -387,6 +388,123 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+        ENVOY_LOG_MISC(trace, "TE header contains a value that exceeds the allowable length");
+        return false;
+      }
+
+      if (is_te_header) {
+        for (const auto& header_value : absl::StrSplit(nominated_header_value_sv, ',')) {
+
+          const absl::string_view header_sv = StringUtil::trim(header_value);
+
+          // If trailers exist in the TE value tokens, keep the header, removing any other values
```
Comment (alyssawilk) [inline source/common/http/utility.cc]: here I assume we don't want to use StringUtil::caseFindToken because it doesn't allow the limitation.
I think that's the right call just because while we could refactor to pass in MAX_ALLOWED_TE_VALUE_SIZE it'd result in complicated return values?
I'd still use StringUtil::splitToken though.

File: source/common/http/utility.cc [hunk 15]
```diff
@@ -387,6 +388,123 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+          if (StringUtil::CaseInsensitiveCompare()(header_sv,
+                                                   Http::Headers::get().TEValues.Trailers)) {
+            keep_header = true;
+            break;
+          }
+        }
+
+        if (keep_header) {
+          nominated_header->value().clear();
+          nominated_header->value().setCopy(Http::Headers::get().TEValues.Trailers.data(),
```
Comment (alyssawilk) [inline source/common/http/utility.cc]: I believe setCopy implicitly clears

File: source/common/http/utility.cc [hunk 16]
```diff
@@ -387,6 +388,123 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
...
+
+  // Lastly remove extra nominated headers from the Connection header
+  if (!headers_to_remove.empty()) {
+    const std::string new_value = StringUtil::removeTokens(connection_header_value.getStringView(),
+                                                           ",", headers_to_remove, ",");
+
+    if (new_value.empty()) {
+      headers.removeConnection();
+    } else {
+      headers.Connection()->value(new_value);
```
Comment (alyssawilk) [inline source/common/http/utility.cc]: does removeTokens remove empty fields?
Maybe a unit test for "foo, ,bar" and removing foo and bar?

File: test/integration/http_integration.cc [hunk 17]
```diff
@@ -337,7 +337,9 @@ void HttpIntegrationTest::verifyResponse(IntegrationStreamDecoderPtr response,
 absl::optional<uint64_t>
 HttpIntegrationTest::waitForNextUpstreamRequest(const std::vector<uint64_t>& upstream_indices,
                                                 std::chrono::milliseconds connection_wait_timeout) {
+
```
Comment (alyssawilk) [inline test/integration/http_integration.cc]: I think you can revert this change?

File: source/common/http/http1/codec_impl.cc (hunk 10 above)
Comment (abaptiste) [reply source/common/http/http1/codec_impl.cc]: Sure thing!

File: test/integration/http_integration.cc (hunk 17 above)
Comment (abaptiste) [reply test/integration/http_integration.cc]: Yep, will clean this up.

File: source/common/http/utility.cc (hunk 15 above)
Comment (abaptiste) [reply source/common/http/utility.cc]: I will remove the `clear` call

File: source/common/http/utility.cc (hunk 13 above)
Comment (abaptiste) [reply source/common/http/utility.cc]: Yep. Added a trace log here

File: docs/root/intro/version_history.rst [hunk 18]
```diff
@@ -3,6 +3,7 @@ Version history
 
 1.13.0 (pending)
 ================
+* http: added the ability to sanitize headers nominated by the Connection header. This new behavior is guarded by envoy.reloadable_features.connection_header_sanitization which defaults to true.
```
Comment (alyssawilk) [inline docs/root/intro/version_history.rst]: mind moving this down?  we try for alphabetic order here

File: source/common/http/utility.cc (hunk 12 above)
Comment (alyssawilk) [reply source/common/http/utility.cc]: Ping on this one?

File: docs/root/intro/version_history.rst (hunk 18 above)
Comment (abaptiste) [reply docs/root/intro/version_history.rst]: Will do.

Comment (abaptiste) [reply docs/root/intro/version_history.rst]: Regarding the X-Forwarded for headers.  I added a comment below the existing one.  I will replace the lines on L445-446 with the updated comment.

See you at EnvoyCon! :)

File: test/integration/header_integration_test.cc [hunk 19]
```diff
@@ -1067,4 +1068,453 @@ TEST_P(HeaderIntegrationTest, TestPathAndRouteOnNormalizedPath) {
           {":status", "200"},
       });
 }
+
+// Validates TE header is forwarded if it contains a supported value
+TEST_P(HeaderIntegrationTest, TestTeHeaderPassthrough) {
```
Comment (alyssawilk) [inline test/integration/header_integration_test.cc]: You know, it only occurs to me now, but we should probably leave one of these tests here to verify connection sanitization works end to end, and do the rest as unit tests of the new function in test/common/http/utility_test.cc
I'm going to say this is 100% optional, since you've done all the work and I forgot to call it out like 8 reviews ago, but if you want to earn alyssawilk brownie points (probably not actually redeemable for brownies) it would be cheaper/faster for CI.

File: docs/root/intro/version_history.rst [hunk 20]
```diff
@@ -11,6 +11,7 @@ Version history
 * health check: gRPC health checker sets the gRPC deadline to the configured timeout duration.
 * http: support :ref:`auto_host_rewrite_header<envoy_api_field_config.filter.http.dynamic_forward_proxy.v2alpha.PerRouteConfig.auto_host_rewrite_header>` in the dynamic forward proxy.
 * logger: added :ref:`--log-format-escaped <operations_cli>` command line option to escape newline characters in application logs.
+* http: added the ability to sanitize headers nominated by the Connection header. This new behavior is guarded by envoy.reloadable_features.connection_header_sanitization which defaults to true.
```
[the remark this answers is not in the document]
Comment (abaptiste) [reply docs/root/intro/version_history.rst]: ugh, vim sort failed me :(  Will fix ... again.

File: test/integration/header_integration_test.cc (hunk 19 above)
Comment (abaptiste) [reply test/integration/header_integration_test.cc]: lol :) It's a deal!!

File: test/common/http/utility_test.cc [hunk 21]
```diff
@@ -754,6 +754,188 @@ TEST(HttpUtility, GetMergedPerFilterConfig) {
...
+      {":authority", "no-headers.com"},
+      {"x-request-foo", "downstram"},
+      {"connection", "te, mike, sam, will, close"},
+      {"te", "gzip, trailers"},
+      {"sam", "bar"},
+      {"will", "baz"},
+  };
+  // Expect that the set of headers is valid and can be sanitized
+  EXPECT_TRUE(Utility::sanitizeConnectionHeader(request_headers));
+}
```
Comment (yanavlasov) [inline test/common/http/utility_test.cc]: My brownie points are not nearly as valuable as Alyssa's. But here a last suggestion I have. I think it will make these tests more informative by also checking the expected header map after sanitization. You can add something like this:
```C++
  TestHeaderMapImpl sanitized_headers = {
      {":method", "GET"},
      {":path", "/"},
      {":scheme", "http"},
      {":authority", "no-headers.com"},
      {"x-request-foo", "downstram"},
      {"connection", "te,close"},
      {"te", "trailers"},
  };
  EXPECT_EQ(sanitized_headers, request_headers);
```
Note that multivalue headers are compared as single value, so no space between "te,close".
I really appreciate your patience here.

Comment (abaptiste) [reply test/common/http/utility_test.cc]: Likewise.  I will update.  Thanks for the suggestion.
